### ISSUE TITLE
Cant connect to production database

### ISSUE THREAD

**Issue by karlclement** [balderdashy/sails-postgresql#211]
I have been at this all day, and found absolutely nothing online to help. When sails lift on my server, I get this for trying to connect to my production database. I tried 3 different postgres DB locations. There is probably something wrong with my settings but I have no idea where to look since the error message gives me nothing. Here is the error:

```
/var/www/url.com/releases/1449515660/node_modules/sails-postgresql/lib/adapter.js:393
        Object.keys(collection.schema).forEach(function(schemaKey) {
               ^
TypeError: Object.keys called on non-object
    at Function.keys (native)
    at __CREATE__ (/var/www/url.com/releases/1449515660/node_modules/sails-postgresql/lib/adapter.js:393:16)
    at after (/var/www/url.com/releases/1449515660/node_modules/sails-postgresql/lib/adapter.js:1208:7)
    at /var/www/url.com/releases/1449515660/node_modules/sails-postgresql/lib/adapter.js:1102:7
    at /var/www/url.com/releases/1449515660/node_modules/sails-postgresql/node_modules/pg/lib/pool.js:82:9
    at /var/www/url.com/releases/1449515660/node_modules/sails-postgresql/node_modules/pg/node_modules/generic-pool/lib/generic-pool.js:281:11
    at /var/www/url.com/releases/1449515660/node_modules/sails-postgresql/node_modules/pg/lib/pool.js:56:18
    at null.<anonymous> (/var/www/url.com/releases/1449515660/node_modules/sails-postgresql/node_modules/pg/lib/client.js:149:7)
    at g (events.js:199:16)
    at emit (events.js:129:20)
    at Socket.<anonymous> (/var/www/url.com/releases/1449515660/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:109:12)
    at Socket.emit (events.js:107:17)
    at readableAddChunk (_stream_readable.js:163:16)
    at Socket.Readable.push (_stream_readable.js:126:10)
    at TCP.onread (net.js:538:20)
```

Here is my connections.js file:
```
devPostgresqlServer: {
    adapter: 'sails-postgresql',
    host: 'localhost',
    user: 'username',
    password: '',
    database: 'databasename',
    ssl: false,
    schema: true
  },

  herokuPostgresqlServer: {
    adapter: 'sails-postgresql',
    host: '**************.amazonaws.com',
    user: '********',
    password: '*********',
    database: '*************',
    ssl: true,
    schema: true
  },

  prodAmazonPostgresqlServer: {
    adapter: 'sails-postgresql',
    host: '*********.amazonaws.com',
    user: '********',
    password: '*************',
    database: '********',
    autoPK: false,
    ssl: false,
    poolSize: 0,
    schema: true
  },

  prodPostgresqlServer: {
    adapter: 'sails-postgresql',
    host: 'localhost',
    user: '**********',
    password: '************',
    database: '***************',
    ssl: false,
    schema: true
  }
```

My local connection works perfectly fine, all 3 other prod versions do not work. PLEASEEEEEE help me!!!! Thanks so much

**Comment by kevinburkeshyp** [comment]
can you console.log `connection.schema` immediately before you get the
TypeError? what does it report?

On Mon, Dec 7, 2015 at 11:53 AM, Karl Clement <notifications@github.com>
wrote:

> I have been at this all day, and found absolutely nothing online to help.
> When sails lift on my server, I get this for trying to connect to my
> production database. I tried 3 different postgres DB locations. There is
> probably something wrong with my settings but I have no idea where to look
> since the error message gives me nothing. Here is the error:
>
> /var/www/weareranger.com/releases/1449515660/node_modules/sails-postgresql/lib/adapter.js:393
>         Object.keys(collection.schema).forEach(function(schemaKey) {
>                ^
> TypeError: Object.keys called on non-object
>     at Function.keys (native)
>     at __CREATE__ (/var/www/weareranger.com/releases/1449515660/node_modules/sails-postgresql/lib/adapter.js:393:16)
>     at after (/var/www/weareranger.com/releases/1449515660/node_modules/sails-postgresql/lib/adapter.js:1208:7)
>     at /var/www/weareranger.com/releases/1449515660/node_modules/sails-postgresql/lib/adapter.js:1102:7
>     at /var/www/weareranger.com/releases/1449515660/node_modules/sails-postgresql/node_modules/pg/lib/pool.js:82:9
>     at /var/www/weareranger.com/releases/1449515660/node_modules/sails-postgresql/node_modules/pg/node_modules/generic-pool/lib/generic-pool.js:281:11
>     at /var/www/weareranger.com/releases/1449515660/node_modules/sails-postgresql/node_modules/pg/lib/pool.js:56:18
>     at null.<anonymous> (/var/www/weareranger.com/releases/1449515660/node_modules/sails-postgresql/node_modules/pg/lib/client.js:149:7)
>     at g (events.js:199:16)
>     at emit (events.js:129:20)
>     at Socket.<anonymous> (/var/www/weareranger.com/releases/1449515660/node_modules/sails-postgresql/node_modules/pg/lib/connection.js:109:12)
>     at Socket.emit (events.js:107:17)
>     at readableAddChunk (_stream_readable.js:163:16)
>     at Socket.Readable.push (_stream_readable.js:126:10)
>     at TCP.onread (net.js:538:20)
>
> Here is my connections.js file:
>
> devPostgresqlServer: {
>     adapter: 'sails-postgresql',
>     host: 'localhost',
>     user: 'username',
>     password: '',
>     database: 'databasename',
>     ssl: false,
>     schema: true
>   },
>
>   herokuPostgresqlServer: {
>     adapter: 'sails-postgresql',
>     host: '**************.amazonaws.com',
>     user: '********',
>     password: '*********',
>     database: '*************',
>     ssl: true,
>     schema: true
>   },
>
>   prodAmazonPostgresqlServer: {
>     adapter: 'sails-postgresql',
>     host: '*********.amazonaws.com',
>     user: '********',
>     password: '*************',
>     database: '********',
>     autoPK: false,
>     ssl: false,
>     poolSize: 0,
>     schema: true
>   },
>
>   prodPostgresqlServer: {
>     adapter: 'sails-postgresql',
>     host: 'localhost',
>     user: 'wearerangerprod',
>     password: '************',
>     database: 'wearerangerprod',
>     ssl: false,
>     schema: true
>   }
>
> My local connection works perfectly fine, all 3 other prod versions do not
> work. PLEASEEEEEE help me!!!! Thanks so much
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balderdashy/sails-postgresql/issues/211>.
>



-- 
kevin

**Comment by karlclement** [comment]
@kevinburkeshyp it returns `undefined`

**Comment by kevinburkeshyp** [comment]
what is the `collection` object? can you trace that, when the connection is
successful vs. unsuccessful?

On Mon, Dec 7, 2015 at 12:29 PM, Karl Clement <notifications@github.com>
wrote:

> @kevinburkeshyp <https://github.com/kevinburkeshyp> it returns undefined
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balderdashy/sails-postgresql/issues/211#issuecomment-162651325>
> .
>



-- 
kevin

**Comment by karlclement** [comment]
@kevinburkeshyp So I went all the way up to console log the err message if it failed to connect and it didn't return any errors. So I decided to console log the collection object and here is what I got:

```
{ connections: { prodPostgresqlServer: { config: [Object], _adapter: [Object], _collections: [Object] } },
  waterline:
   { _collections: [ [Object], [Object], [Object] ],
     _connections: {},
     collections:
      { group: [Object],
        profile: [Object],
        user: [Circular],
        group_users__user_groups: [Object],
        profile_users__user_profiles: [Object] },
     connections:
      { devPostgresqlServer: [Object],
        herokuPostgresqlServer: [Object],
        prodPostgresqlServer: [Object] },
     schema:
      { group: [Object],
        profile: [Object],
        user: [Object],
        group_users__user_groups: [Object],
        profile_users__user_profiles: [Object] } },
  adapter:
   { connections: { prodPostgresqlServer: [Object] },
     dictionary:
      { identity: 'prodPostgresqlServer',
        pkFormat: 'prodPostgresqlServer',
        syncable: 'prodPostgresqlServer',
        defaults: 'prodPostgresqlServer',
        registerConnection: 'prodPostgresqlServer',
        teardown: 'prodPostgresqlServer',
        query: 'prodPostgresqlServer',
        describe: 'prodPostgresqlServer',
        define: 'prodPostgresqlServer',
        drop: 'prodPostgresqlServer',
        addAttribute: 'prodPostgresqlServer',
        removeAttribute: 'prodPostgresqlServer',
        create: 'prodPostgresqlServer',
        createEach: 'prodPostgresqlServer',
        join: 'prodPostgresqlServer',
        find: 'prodPostgresqlServer',
        count: 'prodPostgresqlServer',
        stream: 'prodPostgresqlServer',
        update: 'prodPostgresqlServer',
        destroy: 'prodPostgresqlServer' },
     query: [Circular],
     collection: 'user',
     identity: 'user' },
  _attributes:
   { id:
      { type: 'integer',
        autoIncrement: true,
        primaryKey: true,
        unique: true },
     email: { type: 'email', required: true, unique: true },
     password: { type: 'string', minLength: 6, required: true },
     profiles: { collection: 'profile', via: 'users', dominant: true },
     groups: { collection: 'group', via: 'users' },
     createdAt: { type: 'datetime', default: 'NOW' },
     updatedAt: { type: 'datetime', default: 'NOW' } },
  defaults: { migrate: 'alter', connection: 'prodPostgresqlServer' },
  _cast:
   { _types:
      { id: 'integer',
        email: 'string',
        password: 'string',
        createdAt: 'datetime',
        updatedAt: 'datetime' } },
  _schema:
   { context: [Circular],
     schema:
      { id: [Object],
        email: [Object],
        password: [Object],
        createdAt: [Object],
        updatedAt: [Object] },
     hasSchema: true },
  _validator:
   { validations:
      { id: [Object],
        email: [Object],
        password: [Object],
        profiles: {},
        groups: {} },
     reservedProperties:
      [ 'defaultsTo',
        'primaryKey',
        'autoIncrement',
        'unique',
        'index',
        'collection',
        'dominant',
        'through',
        'columnName',
        'foreignKey',
        'references',
        'on',
        'groupKey',
        'model',
        'via',
        'size',
        'example',
        'validationMessage',
        'validations',
        'populateSettings',
        'onKey',
        'protected' ] },
  _callbacks:
   { beforeValidate: [ [Function] ],
     afterValidate: [ [Function] ],
     beforeUpdate: [ [Function] ],
     afterUpdate: [ [Function] ],
     beforeCreate: [ [Function] ],
     afterCreate: [ [Function] ],
     beforeDestroy: [ [Function] ],
     afterDestroy: [ [Function] ] },
  _instanceMethods: { toJSON: [Function] },
  hasSchema: true,
  migrate: 'safe',
  _model: [Function: bound],
  primaryKey: 'id',
  _transformer: { _transformations: {} },
  adapterDictionary:
   { identity: 'prodPostgresqlServer',
     pkFormat: 'prodPostgresqlServer',
     syncable: 'prodPostgresqlServer',
     defaults: 'prodPostgresqlServer',
     registerConnection: 'prodPostgresqlServer',
     teardown: 'prodPostgresqlServer',
     query: 'prodPostgresqlServer',
     describe: 'prodPostgresqlServer',
     define: 'prodPostgresqlServer',
     drop: 'prodPostgresqlServer',
     addAttribute: 'prodPostgresqlServer',
     removeAttribute: 'prodPostgresqlServer',
     create: 'prodPostgresqlServer',
     createEach: 'prodPostgresqlServer',
     join: 'prodPostgresqlServer',
     find: 'prodPostgresqlServer',
     count: 'prodPostgresqlServer',
     stream: 'prodPostgresqlServer',
     update: 'prodPostgresqlServer',
     destroy: 'prodPostgresqlServer' },
  pkFormat: 'integer',
  syncable: true,
  registerConnection: [Function: bound],
  teardown: [Function: bound],
  query: [Function: bound],
  define: [Function: bound],
  addAttribute: [Function: bound],
  removeAttribute: [Function: bound],
  findOneById: [Function: bound],
  findOneByIdIn: [Function: bound],
  findOneByIdLike: [Function: bound],
  findById: [Function: bound],
  findByIdIn: [Function: bound],
  findByIdLike: [Function: bound],
  countById: [Function: bound],
  countByIdIn: [Function: bound],
  countByIdLike: [Function: bound],
  idStartsWith: [Function: bound],
  idContains: [Function: bound],
  idEndsWith: [Function: bound],
  findOneByEmail: [Function: bound],
  findOneByEmailIn: [Function: bound],
  findOneByEmailLike: [Function: bound],
  findByEmail: [Function: bound],
  findByEmailIn: [Function: bound],
  findByEmailLike: [Function: bound],
  countByEmail: [Function: bound],
  countByEmailIn: [Function: bound],
  countByEmailLike: [Function: bound],
  emailStartsWith: [Function: bound],
  emailContains: [Function: bound],
  emailEndsWith: [Function: bound],
  findOneByPassword: [Function: bound],
  findOneByPasswordIn: [Function: bound],
  findOneByPasswordLike: [Function: bound],
  findByPassword: [Function: bound],
  findByPasswordIn: [Function: bound],
  findByPasswordLike: [Function: bound],
  countByPassword: [Function: bound],
  countByPasswordIn: [Function: bound],
  countByPasswordLike: [Function: bound],
  passwordStartsWith: [Function: bound],
  passwordContains: [Function: bound],
  passwordEndsWith: [Function: bound],
  findOneByCreatedAt: [Function: bound],
  findOneByCreatedAtIn: [Function: bound],
  findOneByCreatedAtLike: [Function: bound],
  findByCreatedAt: [Function: bound],
  findByCreatedAtIn: [Function: bound],
  findByCreatedAtLike: [Function: bound],
  countByCreatedAt: [Function: bound],
  countByCreatedAtIn: [Function: bound],
  countByCreatedAtLike: [Function: bound],
  createdAtStartsWith: [Function: bound],
  createdAtContains: [Function: bound],
  createdAtEndsWith: [Function: bound],
  findOneByUpdatedAt: [Function: bound],
  findOneByUpdatedAtIn: [Function: bound],
  findOneByUpdatedAtLike: [Function: bound],
  findByUpdatedAt: [Function: bound],
  findByUpdatedAtIn: [Function: bound],
  findByUpdatedAtLike: [Function: bound],
  countByUpdatedAt: [Function: bound],
  countByUpdatedAtIn: [Function: bound],
  countByUpdatedAtLike: [Function: bound],
  updatedAtStartsWith: [Function: bound],
  updatedAtContains: [Function: bound],
  updatedAtEndsWith: [Function: bound],
  definition:
   { id:
      { type: 'integer',
        autoIncrement: true,
        primaryKey: true,
        unique: true },
     email: { type: 'string', unique: true },
     password: { type: 'string' },
     createdAt: { type: 'datetime' },
     updatedAt: { type: 'datetime' } },
  meta: { junctionTable: false },
  alter: [Function: bound],
  beforeCreate: [Function: bound],
  buildDynamicFinders: [Function: bound],
  constructor: [Function: bound],
  contains: [Function: bound],
  count: [Function: bound],
  create: [Function: bound],
  createEach: [Function: bound],
  describe: [Function: bound],
  destroy: [Function: bound],
  drop: [Function: bound],
  endsWith: [Function: bound],
  find: [Function: bound],
  findAll: [Function: bound],
  findLike: [Function: bound],
  findOne: [Function: bound],
  findOneLike: [Function: bound],
  findOrCreate: [Function: bound],
  findOrCreateEach: [Function: bound],
  generateAssociationFinders: [Function: bound],
  generateDynamicFinder: [Function: bound],
  join: [Function: bound],
  select: [Function: bound],
  startsWith: [Function: bound],
  stream: [Function: bound],
  sync: [Function: bound],
  update: [Function: bound],
  validate: [Function: bound],
  where: [Function: bound],
  associations:
   [ { alias: 'profiles',
       type: 'collection',
       collection: 'profile',
       via: 'users' },
     { alias: 'groups',
       type: 'collection',
       collection: 'group',
       via: 'users' } ],
  broadcast: [Function],
  getAllContexts: [Function],
  message: [Function],
  publish: [Function: bound],
  pluralize: [Function],
  room: [Function: bound],
  classRoom: [Function],
  _classRoom: [Function],
  subscribers: [Function],
  watchers: [Function],
  subscribe: [Function: bound],
  unsubscribe: [Function: bound],
  publishUpdate: [Function: bound],
  publishDestroy: [Function: bound],
  publishAdd: [Function: bound],
  publishRemove: [Function: bound],
  publishCreate: [Function: bound],
  watch: [Function: bound],
  unwatch: [Function: bound],
  introduce: [Function: bound],
  retire: [Function: bound],
  autosubscribe: true }
```

**Comment by karlclement** [comment]
@kevinburkeshyp any idea what it could be? What else can I do, I'm considering replacing waterline altogether. Thanks again

**Comment by karlclement** [comment]
@particlebanana @kevinburkeshyp  Another thing to note, I don't have any tables on this DB. Do I need to set anything else up to create those tables? They don't get added on sails lift

**Comment by devinivy** [comment]
To create the tables you must start the app _not_ in `production` mode and set your collection migration strategies to `alter` or `drop` (or possibly `create`– not sure if this is supported by sails-postgresql).  These aren't production-viable migration strategies, but they may be used to create your tables.

**Comment by sailsbot** [comment]
Thanks for posting, @karlclement.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by danielgranat** [comment]
@karlclement Have you figured it out? I'm having the same problem.

**Comment by karlclement** [comment]
@danielgranat make sure to set your migration strategy in production to `alter` instead of safe mode. Once it starts sails server, the tables will be created. Then go back and change it back to safe mode to make sure you arent ersasing any production data.

**Comment by danielgranat** [comment]
@karlclement Tried that once on production db and It ended up deleting my users table, as it's loading all the data into memory, drops the table and re-creates it, but there was not enough memory and alter couldn't finish. I use sails-migrations for production migrates.
I found today that as i am using tableName, the name was too long and when i shortened the name it worked :)
